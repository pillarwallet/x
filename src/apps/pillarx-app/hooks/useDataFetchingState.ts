/* eslint-disable @typescript-eslint/no-explicit-any */
import { useEffect } from 'react';

// reducer
import { useAppDispatch } from './useReducerHooks';

export const useDataFetchingState = <T>(
  data: T | undefined,
  isLoading: boolean,
  isFetching: boolean,
  isSuccess: boolean,
  error: any,
  setData: (data: T | undefined) => any,
  setIsLoading: (isLoading: boolean) => any,
  setIsErroring?: (isErroring: boolean) => any
) => {
  const dispatch = useAppDispatch();

  // Update loading state
  useEffect(() => {
    dispatch(setIsLoading(isLoading || isFetching));
  }, [dispatch, isLoading, isFetching, setIsLoading]);

  // Update data and error states
  useEffect(() => {
    const isError = !isLoading && !isFetching && !!error;

    if (data && isSuccess) {
      dispatch(setData(data));
      if (setIsErroring) {
        dispatch(setIsErroring(false));
      }
    } else if (isError) {
      dispatch(setData(undefined));
      if (setIsErroring) {
        dispatch(setIsErroring(true));
      }
    }
  }, [
    dispatch,
    data,
    isSuccess,
    error,
    isLoading,
    isFetching,
    setData,
    setIsErroring,
  ]);
};
