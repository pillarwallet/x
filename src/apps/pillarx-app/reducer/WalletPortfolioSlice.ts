/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { sub } from 'date-fns';

// types
import {
  PortfolioData,
  TokenPriceGraphPeriod,
  WalletHistory,
} from '../../../types/api';

// utils
import { convertDateToUnixTimestamp } from '../../../utils/common';
import { PeriodFilterBalance, PeriodFilterPnl } from '../utils/portfolio';

export type WalletPortfolioState = {
  walletPortfolio: PortfolioData | undefined;
  isWalletPorfolioLoading: boolean;
  isWalletPortfolioErroring: boolean;
  walletPortfolioWithPnl: PortfolioData | undefined;
  isWalletPorfolioWithPnlLoading: boolean;
  isWalletPortfolioWithPnlErroring: boolean;
  priceGraphPeriod: TokenPriceGraphPeriod;
  periodFilter: PeriodFilterBalance;
  periodFilterPnl: PeriodFilterPnl;
  walletHistoryGraph: WalletHistory | undefined;
  isWalletHistoryGraphLoading: boolean;
  isWalletHistoryGraphErroring: boolean;
  topTokenUnrealizedPnL: WalletHistory | undefined;
  isTopTokenUnrealizedPnLLoading: boolean;
  isTopTokenUnrealizedPnLErroring: boolean;
  selectedBalanceOrPnl: 'balance' | 'pnl';
  isRefreshAll: boolean;
  isReceiveModalOpen: boolean;
};

const initialState: WalletPortfolioState = {
  walletPortfolio: undefined,
  isWalletPorfolioLoading: false,
  isWalletPortfolioErroring: false,
  walletPortfolioWithPnl: undefined,
  isWalletPorfolioWithPnlLoading: false,
  isWalletPortfolioWithPnlErroring: false,
  priceGraphPeriod: {
    from: convertDateToUnixTimestamp(sub(new Date(), { days: 1 })),
    to: undefined,
  },
  periodFilter: PeriodFilterBalance.DAY,
  periodFilterPnl: PeriodFilterPnl.DAY,
  walletHistoryGraph: undefined,
  isWalletHistoryGraphLoading: false,
  isWalletHistoryGraphErroring: false,
  topTokenUnrealizedPnL: undefined,
  isTopTokenUnrealizedPnLLoading: false,
  isTopTokenUnrealizedPnLErroring: false,
  selectedBalanceOrPnl: 'balance',
  isRefreshAll: false,
  isReceiveModalOpen: false,
};

const walletPortfolioSlice = createSlice({
  name: 'walletPortfolio',
  initialState,
  reducers: {
    setWalletPortfolio(
      state,
      action: PayloadAction<PortfolioData | undefined>
    ) {
      state.walletPortfolio = action.payload;
    },
    setIsWalletPorfolioLoading(state, action: PayloadAction<boolean>) {
      state.isWalletPorfolioLoading = action.payload;
    },
    setIsWalletPortfolioErroring(state, action: PayloadAction<boolean>) {
      state.isWalletPortfolioErroring = action.payload;
    },
    setWalletPortfolioWithPnl(
      state,
      action: PayloadAction<PortfolioData | undefined>
    ) {
      state.walletPortfolioWithPnl = action.payload;
    },
    setIsWalletPortfolioWithPnlLoading(state, action: PayloadAction<boolean>) {
      state.isWalletPorfolioWithPnlLoading = action.payload;
    },
    setIsWalletPortfolioWithPnlErroring(state, action: PayloadAction<boolean>) {
      state.isWalletPortfolioWithPnlErroring = action.payload;
    },
    setPriceGraphPeriod(state, action: PayloadAction<TokenPriceGraphPeriod>) {
      state.priceGraphPeriod = action.payload;
    },
    setPeriodFilter(state, action: PayloadAction<PeriodFilterBalance>) {
      state.periodFilter = action.payload;
    },
    setPeriodFilterPnl(state, action: PayloadAction<PeriodFilterPnl>) {
      state.periodFilterPnl = action.payload;
    },
    setWalletHistoryGraph(
      state,
      action: PayloadAction<WalletHistory | undefined>
    ) {
      state.walletHistoryGraph = action.payload;
    },
    setIsWalletHistoryGraphLoading(state, action: PayloadAction<boolean>) {
      state.isWalletHistoryGraphLoading = action.payload;
    },
    setIsWalletHistoryGraphErroring(state, action: PayloadAction<boolean>) {
      state.isWalletHistoryGraphErroring = action.payload;
    },
    setTopTokenUnrealizedPnL(
      state,
      action: PayloadAction<WalletHistory | undefined>
    ) {
      state.topTokenUnrealizedPnL = action.payload;
    },
    setIsTopTokenUnrealizedPnLLoading(state, action: PayloadAction<boolean>) {
      state.isTopTokenUnrealizedPnLLoading = action.payload;
    },
    setIsTopTokenUnrealizedPnLErroring(state, action: PayloadAction<boolean>) {
      state.isTopTokenUnrealizedPnLErroring = action.payload;
    },
    setSelectedBalanceOrPnl(state, action: PayloadAction<'balance' | 'pnl'>) {
      state.selectedBalanceOrPnl = action.payload;
    },
    setIsRefreshAll(state, action: PayloadAction<boolean>) {
      state.isRefreshAll = action.payload;
    },
    setIsReceiveModalOpen(state, action: PayloadAction<boolean>) {
      state.isReceiveModalOpen = action.payload;
    },
  },
});

export const {
  setWalletPortfolio,
  setIsWalletPorfolioLoading,
  setIsWalletPortfolioErroring,
  setWalletPortfolioWithPnl,
  setIsWalletPortfolioWithPnlLoading,
  setIsWalletPortfolioWithPnlErroring,
  setPriceGraphPeriod,
  setPeriodFilter,
  setPeriodFilterPnl,
  setWalletHistoryGraph,
  setIsWalletHistoryGraphLoading,
  setIsWalletHistoryGraphErroring,
  setTopTokenUnrealizedPnL,
  setIsTopTokenUnrealizedPnLLoading,
  setIsTopTokenUnrealizedPnLErroring,
  setSelectedBalanceOrPnl,
  setIsRefreshAll,
  setIsReceiveModalOpen,
} = walletPortfolioSlice.actions;

export default walletPortfolioSlice;
