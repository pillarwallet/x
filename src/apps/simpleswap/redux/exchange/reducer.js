import { createReducer } from '@reduxjs/toolkit';

import * as actions from './actions';

const initialState = {
  allCurrenciesList: {
    data: null,
    isLoading: true,
  },
  filter: {
    data: null,
    isLoading: true,
  },
  fixedCurrencies: null,
  allPairsList: null,
  fixedPairs: null,
  fiat: null,
  popularCurrencies: null,
  popularFiat: null,
  exchangesList: [],
};

const reducer = createReducer(initialState, (builder) => {
  builder.addCase(actions.setAllCurrencies, (state, action) => {
    const { data, isLoading } = action.payload;
    if (isLoading)
      return {
        ...state,
        allCurrenciesList: { data: null, isLoading: true },
      };

    const mapCoins = data.filter((coinItem) => !coinItem.isFiat);
    return {
      ...state,
      allCurrenciesList: { data: mapCoins, isLoading: false },
    };
  })
    .addCase(actions.setAllPairs, (state, action) => {
      const { data } = action.payload;

      return {
        ...state,
        allPairsList: data,
      };
    })
    .addCase(actions.setFixedCurrencies, (state, action) => {
      const { data } = action.payload;

      return {
        ...state,
        fixedCurrencies: data,
      };
    })
    .addCase(actions.setFixedPairs, (state, action) => {
      const { data: fixedPairs } = action.payload;

      return {
        ...state,
        fixedPairs,
      };
    })
    .addCase(actions.setPopularCurrencies, (state, action) => {
      const { popularCurrencies } = action.payload;

      return {
        ...state,
        popularCurrencies,
      };
    })
    .addCase(actions.setExchangeInfo, (state, action) => {
      const { data, exchangeId } = action.payload;

      const { exchangesList } = state;

      if (data.error) {
        return state;
      }

      const newExchangesList = exchangesList.filter((i) => i.id !== exchangeId);
      newExchangesList.push(data.data);

      return {
        ...state,
        exchangesList: newExchangesList,
      };
    })
})


export default reducer;
