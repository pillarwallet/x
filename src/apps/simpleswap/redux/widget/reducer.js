/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from '../../helpers/axios';
import {widgetInfo} from '../../constants/widgetInfo';

const initialState = {
  exchangeInfo: null,
};

export const GET_WIDGET_EXCHANGE_INFO = 'widget/GET_WIDGET_EXCHANGE_INFO';

// eslint-disable-next-line consistent-return
export const getWidgetExchangeInfo = createAsyncThunk(
  GET_WIDGET_EXCHANGE_INFO,
  async ({ exchangeId }) => {
      const headers = {
        'x-api-key': widgetInfo.apiKey,
      };
      const response = await axios.get(`/exchanges/${exchangeId}`, {
        headers,
      });

      return response.data.result;
  },
);

export const widgetSlice = createSlice({
  name: 'widget',
  initialState,
  reducers: {
    resetWidgetExchangeInfo: (state) => {
      state.exchangeInfo = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getWidgetExchangeInfo.fulfilled, (state, { payload }) => {
        state.exchangeInfo = payload;
      })
  },
});

export const { resetWidgetExchangeInfo } = widgetSlice.actions;

export default widgetSlice.reducer;
