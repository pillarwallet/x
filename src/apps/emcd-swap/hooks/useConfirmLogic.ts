import { useSelector } from 'react-redux';
import { selectDetailFormDataSwap, selectDetailSwapStatus, selectSwapID } from '../reducer/emcdSwapSlice';

export const useConfirmLogic = () => {
  const detailFormDataSwap = useSelector(selectDetailFormDataSwap)
  const status = useSelector(selectDetailSwapStatus)
  const swapID = useSelector(selectSwapID)

  return {
    detailFormDataSwap,
    swapID,
    status,
  }
}