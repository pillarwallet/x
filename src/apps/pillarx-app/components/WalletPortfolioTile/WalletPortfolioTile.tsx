import { sub } from 'date-fns';
import { useEffect, useMemo } from 'react';

// services
import { useGetWalletHistoryQuery } from '../../../../services/pillarXApiWalletHistory';
import { useGetWalletPortfolioQuery } from '../../../../services/pillarXApiWalletPortfolio';

// types
import { TokenPriceGraphPeriod } from '../../../../types/api';

// utils
import { convertDateToUnixTimestamp } from '../../../../utils/common';
import {
  PeriodFilterBalance,
  getGraphResolutionBalance,
} from '../../utils/portfolio';

// hooks
import useTransactionKit from '../../../../hooks/useTransactionKit';
import { useDataFetchingState } from '../../hooks/useDataFetchingState';

// reducer
import { useAppDispatch, useAppSelector } from '../../hooks/useReducerHooks';
import {
  setIsRefreshAll,
  setIsTopTokenUnrealizedPnLErroring,
  setIsTopTokenUnrealizedPnLLoading,
  setIsWalletHistoryGraphErroring,
  setIsWalletHistoryGraphLoading,
  setIsWalletPortfolioErroring,
  setIsWalletPortfolioLoading,
  setIsWalletPortfolioWithPnlErroring,
  setIsWalletPortfolioWithPnlLoading,
  setTopTokenUnrealizedPnL,
  setWalletHistoryGraph,
  setWalletPortfolio,
  setWalletPortfolioWithPnl,
} from '../../reducer/WalletPortfolioSlice';

// components
import PrimeTokensBalance from '../PrimeTokensBalance/PrimeTokensBalance';
import TileContainer from '../TileContainer/TileContainer';
import TopTokens from '../TopTokens/TopTokens';
import WalletPortfolioBalance from '../WalletPortfolioBalance/WalletPortfolioBalance';
import WalletPortfolioButtons from '../WalletPortfolioButtons/WalletPortfolioButtons';
import WalletPortfolioGraph from '../WalletPortfolioGraph/WalletPortfolioGraph';

const WalletPortfolioTile = () => {
  const { walletAddress: accountAddress } = useTransactionKit();

  const dispatch = useAppDispatch();

  const priceGraphPeriod = useAppSelector(
    (state) => state.walletPortfolio.priceGraphPeriod as TokenPriceGraphPeriod
  );
  const periodFilter = useAppSelector(
    (state) => state.walletPortfolio.periodFilter as PeriodFilterBalance
  );
  const selectedBalanceOrPnl = useAppSelector(
    (state) => state.walletPortfolio.selectedBalanceOrPnl as 'balance' | 'pnl'
  );
  const isRefreshAll = useAppSelector(
    (state) => state.walletPortfolio.isRefreshAll as boolean
  );

  // Query parameters
  const topTokenUnrealizedPnLQueryArgs = useMemo(
    () => ({
      wallet: accountAddress || '',
      period: '1h',
      from: convertDateToUnixTimestamp(sub(new Date(), { days: 1 })),
    }),
    [accountAddress]
  );

  const walletHistoryDataQueryArgs = useMemo(
    () => ({
      wallet: accountAddress || '',
      period: getGraphResolutionBalance(periodFilter),
      from: priceGraphPeriod.from,
    }),
    [accountAddress, periodFilter, priceGraphPeriod.from]
  );

  const walletPortfolioWithPnlArgs = useMemo(
    () => ({
      wallet: accountAddress || '',
      isPnl: true,
    }),
    [accountAddress]
  );

  const shouldFetchPnl = !!accountAddress && selectedBalanceOrPnl === 'pnl';

  // API Queries
  const {
    data: walletPortfolioData,
    isLoading: isWalletPortfolioDataLoading,
    isFetching: isWalletPortfolioDataFetching,
    isSuccess: isWalletPortfolioDataSuccess,
    error: walletPortfolioDataError,
    refetch: refetchWalletPortfolioData,
  } = useGetWalletPortfolioQuery(
    { wallet: accountAddress || '', isPnl: false },
    { skip: !accountAddress }
  );

  const {
    data: walletPortfolioWithPnlData,
    isLoading: isWalletPortfolioDataWithPnlLoading,
    isFetching: isWalletPortfolioDataWithPnlFetching,
    isSuccess: isWalletPortfolioDataWithPnlSuccess,
    error: walletPortfolioDataWithPnlError,
    refetch: refetchWalletPortfolioWithPnlData,
  } = useGetWalletPortfolioQuery(walletPortfolioWithPnlArgs, {
    skip: !shouldFetchPnl,
  });

  const {
    data: walletHistoryData,
    isLoading: isWalletHistoryDataLoading,
    isFetching: isWalletHistoryDataFetching,
    isSuccess: isWalletHistoryDataSuccess,
    error: walletHistoryDataError,
    refetch: refetchWalletHistoryData,
  } = useGetWalletHistoryQuery(walletHistoryDataQueryArgs, {
    skip: !accountAddress,
  });

  const {
    data: topTokenUnrealizedPnLData,
    isLoading: isTopTokenUnrealizedPnLDataLoading,
    isFetching: isTopTokenUnrealizedPnLDataFetching,
    isSuccess: isTopTokenUnrealizedPnLDataSuccess,
    error: topTokenUnrealizedPnLDataError,
    refetch: refetchTopTokenUnrealizedPnLData,
  } = useGetWalletHistoryQuery(topTokenUnrealizedPnLQueryArgs, {
    skip: !accountAddress,
  });

  useDataFetchingState(
    walletPortfolioData?.result?.data,
    isWalletPortfolioDataLoading,
    isWalletPortfolioDataFetching,
    isWalletPortfolioDataSuccess,
    walletPortfolioDataError,
    setWalletPortfolio,
    setIsWalletPortfolioLoading,
    setIsWalletPortfolioErroring
  );

  useDataFetchingState(
    walletPortfolioWithPnlData?.result?.data,
    isWalletPortfolioDataWithPnlLoading,
    isWalletPortfolioDataWithPnlFetching,
    isWalletPortfolioDataWithPnlSuccess,
    walletPortfolioDataWithPnlError,
    setWalletPortfolioWithPnl,
    setIsWalletPortfolioWithPnlLoading,
    setIsWalletPortfolioWithPnlErroring
  );

  useDataFetchingState(
    walletHistoryData?.result?.data,
    isWalletHistoryDataLoading,
    isWalletHistoryDataFetching,
    isWalletHistoryDataSuccess,
    walletHistoryDataError,
    setWalletHistoryGraph,
    setIsWalletHistoryGraphLoading,
    setIsWalletHistoryGraphErroring
  );

  useDataFetchingState(
    topTokenUnrealizedPnLData?.result?.data,
    isTopTokenUnrealizedPnLDataLoading,
    isTopTokenUnrealizedPnLDataFetching,
    isTopTokenUnrealizedPnLDataSuccess,
    topTokenUnrealizedPnLDataError,
    setTopTokenUnrealizedPnL,
    setIsTopTokenUnrealizedPnLLoading,
    setIsTopTokenUnrealizedPnLErroring
  );

  // eslint-disable-next-line consistent-return
  useEffect(() => {
    if (isRefreshAll) {
      // Only refetch if accountAddress exists
      if (accountAddress) {
        refetchWalletPortfolioData();
        refetchWalletHistoryData();
        refetchTopTokenUnrealizedPnLData();

        // Only refetch PnL data if it should be fetched
        if (shouldFetchPnl) {
          refetchWalletPortfolioWithPnlData();
        }
      }

      const timeout = setTimeout(() => {
        dispatch(setIsRefreshAll(false));
      }, 5000);

      return () => clearTimeout(timeout);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isRefreshAll, accountAddress, shouldFetchPnl]);

  return (
    <TileContainer
      id="wallet-portfolio-tile"
      className="desktop:p-9 desktop:gap-9"
    >
      <div className="flex flex-col rounded-xl desktop:border-[1px] desktop:border-lighter_container_grey p-3.5 w-full desktop:gap-4 gap-3">
        <WalletPortfolioBalance />
        <div className="tablet:hidden mobile:hidden desktop:flex w-full">
          <WalletPortfolioButtons />
        </div>
        <PrimeTokensBalance />
        <div className="desktop:hidden mobile:flex tablet:flex rounded-xl w-full">
          <WalletPortfolioGraph />
        </div>
        <div className="desktop:hidden mobile:flex tablet:flex w-full">
          <WalletPortfolioButtons />
        </div>
        <TopTokens />
      </div>
      <div className="tablet:hidden mobile:hidden desktop:flex rounded-xl border-[1px] border-lighter_container_grey w-full">
        <WalletPortfolioGraph />
      </div>
    </TileContainer>
  );
};

export default WalletPortfolioTile;
