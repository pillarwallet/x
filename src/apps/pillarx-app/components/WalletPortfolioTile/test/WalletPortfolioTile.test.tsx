/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, screen } from '@testing-library/react';
import { Mock, vi } from 'vitest';

// servuces
import * as historyHooks from '../../../../../services/pillarXApiWalletHistory';
import * as apiHooks from '../../../../../services/pillarXApiWalletPortfolio';

// hooks
import useTransactionKit from '../../../../../hooks/useTransactionKit';
import * as fetchStateHook from '../../../hooks/useDataFetchingState';

// reducer
import * as reduxHooks from '../../../hooks/useReducerHooks';
import * as sliceActions from '../../../reducer/WalletPortfolioSlice';

// components
import WalletPortfolioTile from '../WalletPortfolioTile';

// Mock child components to isolate the tile itself
vi.mock('../../WalletPortfolioBalance/WalletPortfolioBalance', () => ({
  __esModule: true,
  default: function WalletPortfolioBalance() {
    return <div>WalletPortfolioBalance</div>;
  },
}));
vi.mock('../../WalletPortfolioButtons/WalletPortfolioButtons', () => ({
  __esModule: true,
  default: function WalletPortfolioButtons() {
    return <div>WalletPortfolioButtons</div>;
  },
}));
vi.mock('../../PrimeTokensBalance/PrimeTokensBalance', () => ({
  __esModule: true,
  default: function PrimeTokensBalance() {
    return <div>PrimeTokensBalance</div>;
  },
}));
vi.mock('../../WalletPortfolioGraph/WalletPortfolioGraph', () => ({
  __esModule: true,
  default: function WalletPortfolioGraph() {
    return <div>WalletPortfolioGraph</div>;
  },
}));
vi.mock('../../TopTokens/TopTokens', () => ({
  __esModule: true,
  default: function TopTokens() {
    return <div>TopTokens</div>;
  },
}));

vi.mock('../../../../../hooks/useTransactionKit');

describe('<WalletPortfolioTile />', () => {
  const dispatch = vi.fn();
  const useTransactionKitMock = useTransactionKit as unknown as Mock;

  beforeEach(() => {
    vi.clearAllMocks();

    // Mock Redux hooks
    vi.spyOn(reduxHooks, 'useAppDispatch').mockReturnValue(dispatch);
    vi.spyOn(reduxHooks, 'useAppSelector').mockImplementation(
      (selectorFn: any) =>
        selectorFn({
          walletPortfolio: {
            priceGraphPeriod: { from: 1111 },
            periodFilter: '24h',
            selectedBalanceOrPnl: 'balance',
            isRefreshAll: false,
          },
        })
    );

    useTransactionKitMock.mockReturnValue({
      walletAddress: '0x1234',
      kit: {},
    });

    // Mock useDataFetchingState
    vi.spyOn(fetchStateHook, 'useDataFetchingState').mockImplementation(
      () => {}
    );

    // Mock queries
    vi.spyOn(apiHooks, 'useGetWalletPortfolioQuery').mockImplementation(
      (args: any) => {
        if (args.isPnl) {
          return {
            data: { result: { data: {} } },
            isLoading: false,
            isFetching: false,
            isSuccess: true,
            error: null,
            refetch: vi.fn(),
          };
        }
        return {
          data: { result: { data: {} } },
          isLoading: false,
          isFetching: false,
          isSuccess: true,
          error: null,
          refetch: vi.fn(),
        };
      }
    );

    vi.spyOn(historyHooks, 'useGetWalletHistoryQuery').mockReturnValue({
      data: { result: { data: {} } },
      isLoading: false,
      isFetching: false,
      isSuccess: true,
      error: null,
      refetch: vi.fn(),
    });
  });

  it('renders correctly and matches snapshot', () => {
    const tree = render(<WalletPortfolioTile />);
    expect(tree).toMatchSnapshot();
  });

  it('renders all tile sections correctly', () => {
    render(<WalletPortfolioTile />);

    expect(screen.getByText('WalletPortfolioBalance')).toBeInTheDocument();
    expect(screen.getAllByText('WalletPortfolioButtons')).toHaveLength(2);
    expect(screen.getByText('PrimeTokensBalance')).toBeInTheDocument();
    expect(screen.getAllByText('WalletPortfolioGraph')).toHaveLength(2);
    expect(screen.getByText('TopTokens')).toBeInTheDocument();
  });

  it('skips rendering parts if no wallet address', () => {
    useTransactionKitMock.mockReturnValue({
      walletAddress: undefined,
      kit: {},
    });

    render(<WalletPortfolioTile />);

    expect(screen.getByText('WalletPortfolioBalance')).toBeInTheDocument();
  });

  it('triggers refetch on isRefreshAll true and dispatch reset after timeout', () => {
    vi.useFakeTimers();

    const refetchMock = vi.fn();
    vi.spyOn(reduxHooks, 'useAppSelector').mockImplementation(
      (selectorFn: any) =>
        selectorFn({
          walletPortfolio: {
            priceGraphPeriod: { from: 1111 },
            periodFilter: '24h',
            selectedBalanceOrPnl: 'pnl',
            isRefreshAll: true,
          },
        })
    );

    vi.spyOn(apiHooks, 'useGetWalletPortfolioQuery').mockImplementation(() => {
      return {
        data: { result: { data: {} } },
        isLoading: false,
        isFetching: false,
        isSuccess: true,
        error: null,
        refetch: refetchMock,
      };
    });

    vi.spyOn(historyHooks, 'useGetWalletHistoryQuery').mockReturnValue({
      data: { result: { data: {} } },
      isLoading: false,
      isFetching: false,
      isSuccess: true,
      error: null,
      refetch: refetchMock,
    });

    render(<WalletPortfolioTile />);
    expect(refetchMock).toHaveBeenCalledTimes(4);

    vi.advanceTimersByTime(5000);
    expect(dispatch).toHaveBeenCalledWith(sliceActions.setIsRefreshAll(false));

    vi.useRealTimers();
  });
});
