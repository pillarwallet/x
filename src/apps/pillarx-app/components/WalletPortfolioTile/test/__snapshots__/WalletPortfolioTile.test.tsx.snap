// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<WalletPortfolioTile /> > renders correctly and matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex bg-container_grey rounded-3xl desktop:p-9 desktop:gap-9"
        id="wallet-portfolio-tile"
      >
        <div
          class="flex flex-col rounded-xl desktop:border-[1px] desktop:border-lighter_container_grey p-3.5 w-full desktop:gap-4 gap-3"
        >
          <div>
            WalletPortfolioBalance
          </div>
          <div
            class="tablet:hidden mobile:hidden desktop:flex w-full"
          >
            <div>
              WalletPortfolioButtons
            </div>
          </div>
          <div>
            PrimeTokensBalance
          </div>
          <div
            class="desktop:hidden mobile:flex tablet:flex rounded-xl w-full"
          >
            <div>
              WalletPortfolioGraph
            </div>
          </div>
          <div
            class="desktop:hidden mobile:flex tablet:flex w-full"
          >
            <div>
              WalletPortfolioButtons
            </div>
          </div>
          <div>
            TopTokens
          </div>
        </div>
        <div
          class="tablet:hidden mobile:hidden desktop:flex rounded-xl border-[1px] border-lighter_container_grey w-full"
        >
          <div>
            WalletPortfolioGraph
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex bg-container_grey rounded-3xl desktop:p-9 desktop:gap-9"
      id="wallet-portfolio-tile"
    >
      <div
        class="flex flex-col rounded-xl desktop:border-[1px] desktop:border-lighter_container_grey p-3.5 w-full desktop:gap-4 gap-3"
      >
        <div>
          WalletPortfolioBalance
        </div>
        <div
          class="tablet:hidden mobile:hidden desktop:flex w-full"
        >
          <div>
            WalletPortfolioButtons
          </div>
        </div>
        <div>
          PrimeTokensBalance
        </div>
        <div
          class="desktop:hidden mobile:flex tablet:flex rounded-xl w-full"
        >
          <div>
            WalletPortfolioGraph
          </div>
        </div>
        <div
          class="desktop:hidden mobile:flex tablet:flex w-full"
        >
          <div>
            WalletPortfolioButtons
          </div>
        </div>
        <div>
          TopTokens
        </div>
      </div>
      <div
        class="tablet:hidden mobile:hidden desktop:flex rounded-xl border-[1px] border-lighter_container_grey w-full"
      >
        <div>
          WalletPortfolioGraph
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<WalletPortfolioTile /> renders correctly and matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex bg-container_grey rounded-3xl desktop:p-9 desktop:gap-9"
        id="wallet-portfolio-tile"
      >
        <div
          class="flex flex-col rounded-xl desktop:border-[1px] desktop:border-lighter_container_grey p-3.5 w-full desktop:gap-4 gap-3"
        >
          <div>
            WalletPortfolioBalance
          </div>
          <div
            class="tablet:hidden mobile:hidden desktop:flex w-full"
          >
            <div>
              WalletPortfolioButtons
            </div>
          </div>
          <div>
            PrimeTokensBalance
          </div>
          <div
            class="desktop:hidden mobile:flex tablet:flex rounded-xl w-full"
          >
            <div>
              WalletPortfolioGraph
            </div>
          </div>
          <div
            class="desktop:hidden mobile:flex tablet:flex w-full"
          >
            <div>
              WalletPortfolioButtons
            </div>
          </div>
          <div>
            TopTokens
          </div>
        </div>
        <div
          class="tablet:hidden mobile:hidden desktop:flex rounded-xl border-[1px] border-lighter_container_grey w-full"
        >
          <div>
            WalletPortfolioGraph
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex bg-container_grey rounded-3xl desktop:p-9 desktop:gap-9"
      id="wallet-portfolio-tile"
    >
      <div
        class="flex flex-col rounded-xl desktop:border-[1px] desktop:border-lighter_container_grey p-3.5 w-full desktop:gap-4 gap-3"
      >
        <div>
          WalletPortfolioBalance
        </div>
        <div
          class="tablet:hidden mobile:hidden desktop:flex w-full"
        >
          <div>
            WalletPortfolioButtons
          </div>
        </div>
        <div>
          PrimeTokensBalance
        </div>
        <div
          class="desktop:hidden mobile:flex tablet:flex rounded-xl w-full"
        >
          <div>
            WalletPortfolioGraph
          </div>
        </div>
        <div
          class="desktop:hidden mobile:flex tablet:flex w-full"
        >
          <div>
            WalletPortfolioButtons
          </div>
        </div>
        <div>
          TopTokens
        </div>
      </div>
      <div
        class="tablet:hidden mobile:hidden desktop:flex rounded-xl border-[1px] border-lighter_container_grey w-full"
      >
        <div>
          WalletPortfolioGraph
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
